code/hsec-sync/src/Security/Advisories/Sync/Snapshot.hs

```diff
@@ -0,0 +1,214 @@
+            ExtractSnapshotArchive x -> "Extraction got an exception: " <> displayException x
+        ]
+
+newtype SnapshotUrl = SnapshotUrl {getSnapshotUrl :: String}
+
+data SnapshotRepositoryStatus
+  = SnapshotDirectoryMissing
+  | SnapshotDirectoryInitialized
+  | SnapshotDirectoryIncoherent
+
+snapshotRepositoryStatus :: FilePath -> IO SnapshotRepositoryStatus
+snapshotRepositoryStatus root = do
+  dirExists <- D.doesDirectoryExist root
+  if dirExists
+    then do
+      dirAdvisoriesExists <- D.doesDirectoryExist $ root </> "advisories"
+      etagMetadataExists <- D.doesFileExist $ root </> "snapshot-etag"
+      return $
+        if dirAdvisoriesExists && etagMetadataExists
+          then SnapshotDirectoryInitialized
+          else SnapshotDirectoryIncoherent
+    else return SnapshotDirectoryMissing
+
+data SnapshotRepositoryEnsuredStatus
+  = SnapshotRepositoryCreated
+  | SnapshotRepositoryExisting
+
+ensureSnapshot ::
+  FilePath ->
+  SnapshotUrl ->
+  SnapshotRepositoryStatus ->
+  ExceptT SnapshotError IO SnapshotRepositoryEnsuredStatus
+ensureSnapshot root url =
+  \case
+    SnapshotDirectoryMissing -> do
+      overwriteSnapshot root url
+      return SnapshotRepositoryCreated
+    SnapshotDirectoryIncoherent -> do
+      overwriteSnapshot root url
+      return SnapshotRepositoryCreated
+    SnapshotDirectoryInitialized ->
+      return SnapshotRepositoryExisting
+
+overwriteSnapshot :: FilePath -> SnapshotUrl -> ExceptT SnapshotError IO ()
+overwriteSnapshot root url =
+  withExceptT SnapshotProcessError $ do
+    ensuringPerformed <- liftIO $ try $ ensureEmptyRoot root
+    whenLeft ensuringPerformed $
+      throwE . DirectorySetupSnapshotArchive
+
+    resultE <- liftIO $ try $ get $ getSnapshotUrl url
+    case resultE of
+      Left e ->
+        throwE $
+          FetchSnapshotArchive $
+            case e of
+              InvalidUrlException url' reason ->
+                "Invalid URL " <> show url' <> ": " <> show reason
+              HttpExceptionRequest _ content ->
+                case content of
+                  StatusCodeException response body ->
+                    "Request failed with " <> show (response ^. responseStatus) <> ": " <> show body
+                  _ ->
+                    "Request failed: " <> show content
+      Right result -> do
+        performed <-
+          liftIO $
+            try $
+              withSystemTempDirectory "security-advisories" $ \tempDir -> do
+                let archivePath = tempDir <> "/snapshot-export.tar.gz"
+                BL.writeFile archivePath $ result ^. responseBody
+                contents <- BL.readFile archivePath
+                let fixEntry e = e { Tar.entryTarPath = fixEntryPath $ Tar.entryTarPath e }
+                    fixEntryPath :: Tar.TarPath -> Tar.TarPath
+                    fixEntryPath p =
+                      fromRight p $
+                        maybe
+                          (Right p)
```

Comment:
This causes the top-level directory/ies and other top-level paths to be created.  It does work *for our use case*, so I'll merge it *as is*.  But I will probably circle back later to tidy it up.